src/parser.ts

```diff
@@ -1326,7 +1326,43 @@ export class Parser {
             } else if (this.lookahead.type === Token.Template && this.lookahead.head) {
                 const quasi = this.parseTemplateLiteral();
                 expr = this.finalize(this.startNode(startToken), new Node.TaggedTemplateExpression(expr, quasi));
-
+            } else if (this.match('?.')) {
+                this.nextToken();
+                if (this.match('(')) {
+                    var asyncArrow = maybeAsync && (startToken.lineNumber === this.lookahead.lineNumber);
```

Comment:
each of these conditions is pretty much copy and pasted from the respective parsing of the non-optional chaining versions of these.

The exception is the static member expression case, we can't use the `parseIdentifierName` function b/c it expects the token to still be at the `.` and handles advancing to the identifier, but we had to advance past to see which type of optional chaining this is.